homeassistant/components/mqtt/util.py

```diff
@@ -216,6 +217,7 @@ def valid_birth_will(config: ConfigType) -> ConfigType:
     return config
 
 
+@lru_cache(maxsize=1)
 def get_mqtt_data(hass: HomeAssistant) -> MqttData:
```

Comment:
Should this function even exist or should we just use a `HassKey` ?

Reply:
At some point it should be migrated to `HassKey`.... but that change was a bit invasive (see below) and considering how many other changes need to happen to solve the event loop being blocked because of the performance issues (https://github.com/home-assistant/core/pull/109030#issuecomment-2123667616), it seemed better to wait until the other issues are sorted out first.

```python
homeassistant/components/mqtt/__init__.py:    get_mqtt_data,
homeassistant/components/mqtt/__init__.py:    mqtt_data = get_mqtt_data(hass)
homeassistant/components/mqtt/__init__.py:            mqtt_data = get_mqtt_data(hass)
homeassistant/components/mqtt/__init__.py:            get_mqtt_data.cache_clear()
homeassistant/components/mqtt/__init__.py:    mqtt_data = get_mqtt_data(hass)
homeassistant/components/mqtt/__init__.py:    mqtt_data = get_mqtt_data(hass)
homeassistant/components/mqtt/client.py:from .util import get_file_path, get_mqtt_data, mqtt_config_entry_enabled
homeassistant/components/mqtt/client.py:    mqtt_data = get_mqtt_data(hass)
homeassistant/components/mqtt/client.py:        mqtt_data = get_mqtt_data(hass)
homeassistant/components/mqtt/debug_info.py:from .util import get_mqtt_data
homeassistant/components/mqtt/debug_info.py:    debug_info_entities = get_mqtt_data(hass).debug_info_entities
homeassistant/components/mqtt/debug_info.py:    entity_info = get_mqtt_data(hass).debug_info_entities.setdefault(
homeassistant/components/mqtt/debug_info.py:        entity_info = get_mqtt_data(hass).debug_info_entities.setdefault(
homeassistant/components/mqtt/debug_info.py:        debug_info_entities := get_mqtt_data(hass).debug_info_entities
homeassistant/components/mqtt/debug_info.py:    entity_info = get_mqtt_data(hass).debug_info_entities.setdefault(
homeassistant/components/mqtt/debug_info.py:    discovery_data = get_mqtt_data(hass).debug_info_entities[entity_id][
homeassistant/components/mqtt/__init__.py:    get_mqtt_data,
homeassistant/components/mqtt/__init__.py:    mqtt_data = get_mqtt_data(hass)
homeassistant/components/mqtt/__init__.py:            mqtt_data = get_mqtt_data(hass)
homeassistant/components/mqtt/__init__.py:            get_mqtt_data.cache_clear()
homeassistant/components/mqtt/__init__.py:    mqtt_data = get_mqtt_data(hass)
homeassistant/components/mqtt/__init__.py:    mqtt_data = get_mqtt_data(hass)
homeassistant/components/mqtt/client.py:from .util import get_file_path, get_mqtt_data, mqtt_config_entry_enabled
homeassistant/components/mqtt/client.py:    mqtt_data = get_mqtt_data(hass)
homeassistant/components/mqtt/client.py:        mqtt_data = get_mqtt_data(hass)
homeassistant/components/mqtt/debug_info.py:from .util import get_mqtt_data
homeassistant/components/mqtt/debug_info.py:    debug_info_entities = get_mqtt_data(hass).debug_info_entities
homeassistant/components/mqtt/debug_info.py:    entity_info = get_mqtt_data(hass).debug_info_entities.setdefault(
homeassistant/components/mqtt/debug_info.py:        entity_info = get_mqtt_data(hass).debug_info_entities.setdefault(
homeassistant/components/mqtt/debug_info.py:        debug_info_entities := get_mqtt_data(hass).debug_info_entities
homeassistant/components/mqtt/debug_info.py:    entity_info = get_mqtt_data(hass).debug_info_entities.setdefault(
homeassistant/components/mqtt/debug_info.py:    discovery_data = get_mqtt_data(hass).debug_info_entities[entity_id][
homeassistant/components/mqtt/debug_info.py:    if entity_id in (debug_info_entities := get_mqtt_data(hass).debug_info_entities):
homeassistant/components/mqtt/debug_info.py:    get_mqtt_data(hass).debug_info_triggers[discovery_hash] = {
homeassistant/components/mqtt/debug_info.py:    get_mqtt_data(hass).debug_info_triggers[discovery_hash]["discovery_data"][
homeassistant/components/mqtt/debug_info.py:    get_mqtt_data(hass).debug_info_triggers.pop(discovery_hash)
homeassistant/components/mqtt/debug_info.py:    entity_info = get_mqtt_data(hass).debug_info_entities[entity_id]
homeassistant/components/mqtt/debug_info.py:    trigger = get_mqtt_data(hass).debug_info_triggers[trigger_key]
homeassistant/components/mqtt/debug_info.py:    mqtt_data = get_mqtt_data(hass)
homeassistant/components/mqtt/debug_info.py:    mqtt_data = get_mqtt_data(hass)
homeassistant/components/mqtt/device_trigger.py:from .util import get_mqtt_data
homeassistant/components/mqtt/device_trigger.py:        self._mqtt_data = get_mqtt_data(hass)
homeassistant/components/mqtt/device_trigger.py:            mqtt_data = get_mqtt_data(self.hass)
homeassistant/components/mqtt/device_trigger.py:    mqtt_data = get_mqtt_data(hass)
homeassistant/components/mqtt/device_trigger.py:    mqtt_data = get_mqtt_data(hass)
homeassistant/components/mqtt/device_trigger.py:    mqtt_data = get_mqtt_data(hass)
homeassistant/components/mqtt/device_trigger.py:    mqtt_data = get_mqtt_data(hass)
homeassistant/components/mqtt/diagnostics.py:from .util import get_mqtt_data
homeassistant/components/mqtt/diagnostics.py:    mqtt_instance = get_mqtt_data(hass).client
homeassistant/components/mqtt/discovery.py:from .util import async_forward_entry_setup_and_setup_discovery, get_mqtt_data
homeassistant/components/mqtt/discovery.py:    get_mqtt_data(hass).discovery_already_discovered.remove(discovery_hash)
homeassistant/components/mqtt/discovery.py:    get_mqtt_data(hass).discovery_already_discovered.add(discovery_hash)
homeassistant/components/mqtt/discovery.py:    mqtt_data = get_mqtt_data(hass)
homeassistant/components/mqtt/discovery.py:    mqtt_data = get_mqtt_data(hass)
homeassistant/components/mqtt/discovery.py:    mqtt_data = get_mqtt_data(hass)
homeassistant/components/mqtt/event.py:from .util import get_mqtt_data
homeassistant/components/mqtt/event.py:            mqtt_data = get_mqtt_data(self.hass)
homeassistant/components/mqtt/image.py:from .util import get_mqtt_data, valid_subscribe_topic
homeassistant/components/mqtt/image.py:            get_mqtt_data(self.hass).state_write_requests.write_state_request(self)
homeassistant/components/mqtt/image.py:            get_mqtt_data(self.hass).state_write_requests.write_state_request(self)
homeassistant/components/mqtt/mixins.py:from .util import get_mqtt_data, mqtt_config_entry_enabled, valid_subscribe_topic
homeassistant/components/mqtt/mixins.py:    mqtt_data = get_mqtt_data(hass)
homeassistant/components/mqtt/mixins.py:    mqtt_data = get_mqtt_data(hass)
homeassistant/components/mqtt/mixins.py:        mqtt_data = get_mqtt_data(hass)
homeassistant/components/mqtt/mixins.py:            mqtt_data = get_mqtt_data(entity.hass)
homeassistant/components/mqtt/mixins.py:        mqtt_data = get_mqtt_data(self.hass)
homeassistant/components/mqtt/mixins.py:        mqtt_data = get_mqtt_data(hass)
homeassistant/components/mqtt/tag.py:from .util import get_mqtt_data, valid_subscribe_topic
homeassistant/components/mqtt/tag.py:    if device_id is not None and device_id not in (tags := get_mqtt_data(hass).tags):
homeassistant/components/mqtt/tag.py:    if device_id not in (tags := get_mqtt_data(hass).tags):
homeassistant/components/mqtt/tag.py:            get_mqtt_data(self.hass).tags[self.device_id].pop(discovery_id)
homeassistant/components/mqtt/util.py:    mqtt_data = get_mqtt_data(hass)
homeassistant/components/mqtt/util.py:def get_mqtt_data(hass: HomeAssistant) -> MqttData:
```